src/matrix_validator/validator_polars.py

```diff
@@ -22,152 +36,140 @@ def __init__(self):
+        validation_reports = []
+
+        bm_prefixes = self.get_biolink_model_prefix_keys()
+        bm_knowledge_level_keys = self.get_biolink_model_knowledge_level_keys()
+        bm_agent_type_keys = self.get_biolink_model_agent_type_keys()
+
         if nodes_file_path:
-            validate_kg_nodes(nodes_file_path, self.output_format, self.get_report_file())
+            validation_reports.extend(validate_kg_nodes(bm_prefixes, nodes_file_path))
 
-        if edges_file_file_path:
-            validate_kg_edges(edges_file_file_path, self.output_format, self.get_report_file())
+        if edges_file_path:
+            validation_reports.extend(validate_kg_edges(bm_prefixes, bm_knowledge_level_keys, bm_agent_type_keys, edges_file_path))
 
+        if nodes_file_path and edges_file_path:
+            validation_reports.extend(validate_nodes_and_edges(nodes_file_path, edges_file_path))
+
+        # Write validation report
+        self.write_report(validation_reports)
+        logging.info(f"Validation report written to {self.get_report_file()}")
+
+
+def validate_nodes_and_edges(nodes, edges):
+    """Validate a knowledge graph nodes vs edges."""
+    logger.info("Validating nodes & edges")
+
+    edges_df = (
+        pl.scan_csv(edges, separator="\t", truncate_ragged_lines=True, has_header=True, ignore_errors=True)
+        .select([pl.col("subject"), pl.col("object")])
+        .collect()
+    )
+    edge_ids = pl.concat([edges_df.select(pl.col("subject").alias("id")), edges_df.select(pl.col("object").alias("id"))]).get_column("id")
+
+    counts_df = (
+        pl.scan_csv(nodes, separator="\t", truncate_ragged_lines=True, has_header=True, ignore_errors=True)
+        .select([(~pl.col("id").str.contains_any(edge_ids)).sum().alias("invalid_edge_ids_in_node_ids_count")])
+        .collect()
+    )
 
-def validate_kg_nodes(nodes, output_format, report_file):
-    """Validate a knowledge graph using optional nodes TSV files."""
     validation_reports = []
 
+    if counts_df.get_column("invalid_edge_ids_in_node_ids_count").item(0) > 0:
+        validation_reports.append(check_edge_ids_in_node_ids(edge_ids, nodes))
+
+    return validation_reports
+
+
+def validate_kg_nodes(bm_prefixes, nodes):
+    """Validate a knowledge graph using optional nodes TSV files."""
     logger.info("Validating nodes TSV...")
 
     counts_df = (
         pl.scan_csv(nodes, separator="\t", truncate_ragged_lines=True, has_header=True, ignore_errors=True)
         .select(
             [
-                pl.col("id").str.contains(CURIE_REGEX).sum().alias("valid_curie_id_count"),
                 (~pl.col("id").str.contains(CURIE_REGEX)).sum().alias("invalid_curie_id_count"),
-                pl.col("category").str.contains(STARTS_WITH_BIOLINK_REGEX).sum().alias("valid_starts_with_biolink_category_count"),
+                (~pl.col("id").str.contains_any(bm_prefixes)).sum().alias("invalid_contains_biolink_model_prefix_id_count"),
```

Comment:
This does not fail gracefully if the the column "id" does not exist.. Should there be a quick inspections of the columns upfront to make sure all the relevant ones are present?

Reply:
I added schema checks.